### PR TITLE
RFC: move mpsse helpers to the separate ftdi-mpsse crate

### PR DESCRIPTION
**Pull request by geomatsi**
Hi @newAM 

As a follow-up to our [discussion](https://github.com/newAM/ftd2xx-embedded-hal/issues/34), this PR proposes to move all the MPSSE helpers and definitions into the separate ```ftdi-mpsse``` crate. The purpose is to allow reuse of MPSSE definitions and command builder outside ```libftd2xx-rs```, e.g. in [ftdi-rs](https://github.com/tanriol/ftdi-rs) and [ftdi-embedded-hal](https://github.com/geomatsi/ftdi-embedded-hal).

Notes:
* new crate has not yet been released, so git repository is temporarily specified in _Cargo.toml_ 
* for the time being the new crate is published here: [ftdi-mpsse](https://github.com/geomatsi/ftdi-mpsse) 
* _common-ftdimpsse_ branch of ```ftdi-embedded-hal``` uses ```ftdi-mpsse```, examples have been modified to support both ```libftd2xx-rs``` and ```ftdi-rs```

I would suggest to move new crate to your github space in the case if proposed approach looks reasonable to you. Well, obviously after all the review comments are addressed. 

Thoughts ? Comments ?

Regards,
Sergey

### CODE REVIEW

**Comment by newAM** [comment]
Thanks for the PR!  I am pretty busy right now with work, I will review it this weekend.

File: src/mpsse.rs
Comment (newAM) [inline src/mpsse.rs:124]: What was the purpose for the rename?
Ideally breaking changes can be kept to a minimum, but I am happy to rename it if there is a good reason.

File: src/lib.rs
Comment (newAM) [inline src/lib.rs:105]: These types should be re-exported from the mpsse crate where applicable, to reduce the number of breaking changes.

**Review by newAM — CHANGES_REQUESTED** [review]
Looks good so far!  Thank you for your work on this!

I do want to wait for a demo `libftdi` implementation before merging to ensure this will work out as intended for both drivers.

File: src/mpsse.rs
Comment (geomatsi) [reply src/mpsse.rs:386]: Yes, almost all those options are suitable for both ```libftd2xx-rs``` and ```ftdi-rs```. Apparently ```libftdi``` implementation of low level hardware operations resembles that of ```libftd2xx```. For instance, here is a simplified version of the current ftdi-rs implementation of mpsse_init:

```
        let millis = u8::try_from(settings.latency_timer.as_millis());

        self.set_write_chunksize(settings.in_transfer_size);
        self.set_read_chunksize(settings.in_transfer_size);
        self.usb_reset();
        self.set_latency_timer(millis);
        self.set_bitmode(0, BitMode::Mpsse);
        self.usb_purge_buffers();

        if let Some(frequency) = settings.clock_frequency {
            self.set_mpsse_clock(frequency);
        }     
```
Missing bits, e.g. read/write timeouts, are implemented by ```libftdi```, but not yet exposed by ```ftdi-rs```.

In fact I have a working WIP implementation for both ```libftd2xx-rs``` and ```ftdi-rs``` drivers in ```ftdi-embedded-hal```, see the following branches:
- [common-mpsse](https://github.com/geomatsi/ftdi-embedded-hal/pull/14) branch for ftdi-embedded-hal
- [common-mpsse](https://github.com/geomatsi/ftdi-rs/tree/common-ftdimpsse)  branch for ftdi-rs

All ```ftdi-embedded-hal``` examples on common-mpsse branch support both ```liftd2xx-rs``` and ```ftdi-rs```. For now I am using 'feature' switch only for examples until I find a better way.

I have not yet submitted PR to ```ftdi-rs``` project. At first I plan to work on your review comments in order to finalize splitting out ```ftdi-mpsse``` crate.

File: src/lib.rs
Comment (geomatsi) [reply src/lib.rs:105]: Fixed: pushed new commit.

File: src/mpsse.rs
Comment (geomatsi) [reply src/mpsse.rs]: No real purpose. So reverting it here and in ftdi-mpsse crate. Fixed: new commit added to this PR.

File: src/mpsse.rs
Comment (geomatsi) [reply src/mpsse.rs]: New commit added to PR: switched to mpsse macros for simple commands. Also I will take a look at gpio read commands: this change should be applicable for them as well.

File: src/lib.rs
Comment (geomatsi) [reply src/lib.rs:2255]: 1. Concerning read_all/write_all
Could you please clarify what you mean ?
Am I correct assuming that you suggest to define MpsseCmdExecutor for std::io::{Read,Write} implementors e.g. along the following lines:
```
pub trait MpsseCmdExecutor: std::io::Read + std::io::Write {
    /// Error type
    type Error;

    /// Configure FTDI MPSSE mode
    fn mpsse_init(&mut self, settings: &MpsseSettings) -> Result<(), Self::Error>;

    /// Execute MPSSE command sequence
    fn mpsse_send(&mut self, data: &[u8]) -> Result<(), Self::Error> {
        self.write_all(data)
    }

    /// Execute MPSSE command and read response
    fn mpsse_xfer(&mut self, txdata: &[u8], rxdata: &mut [u8]) -> Result<(), Self::Error> {
        self.write_all(txdata)?;
        self.read_all(rxdata)
    }

    /// Close FTDI MPSSE device
    fn mpsse_close(&mut self);
}
```

2. On closing mpsse
This is a very good question. In ```ftdi-embedded-hal``` I create HAL handler from device handler that should implement MpsseCmdExecutor. I pass that device handler to HAL by value rather than by reference in order to avoid too many lifetime annotations internally. So that device is consumed by HAL. However after I am done with that HAL instance and  e.g. it leaves the scope, I still can not create another ```libftd2xx-rs``` device instance in the same process for the same device unless I explicitly close previous device instance. This problem does not exist for ```ftdi-rs``` device handlers. Probably  ```libftd2xx``` C library tracks per-process state internally. I tried to implement Drop in ```libftd2xx-rs```, but that approach does not work well with passing device handler into HAL by value because device handler is re-created at some point loosing valid FT_HANDLE value. So for now I use this ```close``` method. Any suggestions are welcome.

3. FtdiMpsse methods
See my comment for 1. Though I don't quite understand how to define default method for mpsse_init and mpsse_close.

Comment (geomatsi) [reply src/lib.rs:2255]: I figured out that my explanation about ```mpsse_close``` is a bit fuzzy. So here are some details and clarifications. I will start from a simple test program:

```rust
use libftd2xx::{Ft232h, FtdiCommon};
use std::time::Duration;

fn main() {
    let ft = Ft232h::with_description("Single RS232-HS").unwrap();
    test(ft, "T1");

    let ft = Ft232h::with_description("Single RS232-HS").unwrap();
    test(ft, "T2");
}   

fn test(mut d: Ft232h, t: &str) {
    println!("{}: TP#1", t);
    d.set_latency_timer(Duration::from_millis(10)).unwrap();
    //d.close().unwrap();
    println!("{}: TP#2", t);
}
```

The second ```with_description``` call fails with ```DEVICE_NOT_OPENED``` if device is not properly closed. Removing comment from the ```close``` call makes things work.

Another option would be to implement ```Drop``` trait. The question is which type should implement it. If ```Drop``` trait is implemented for ```Ftdi```, then it will be called in ```try_from``` when ```Ftdi``` device is promoted to ```Ft232h```. As a result, the above code will fail with ```INVALID_HANDLE``` right on the first ```with_description``` call. Implementing ```Drop``` trait for ```Ftx232h``` types fixes this simple code. However it does not look consistent in general. So I added ```mpsse_close``` method as a temporary fix until proper solution is found.

Maybe ```Ftx232h``` types should embed ```Ftdi``` handler rather than drop it after copying internal handler ?

Comment (geomatsi) [reply src/lib.rs]: New RFC commit has been added to MR: Ftdi embedding and Drop implementation. Now the above example works fine w/o additional close. So mpsse_close is not needed anymore.

Comment (newAM) [reply src/lib.rs]: > Concerning read_all/write_all Could you please clarify what you mean ?

Sorry, I was not very clear.
The method `mpsse_xfer` could be derived with a provided `read_all` and `write_all` function; should the methods that are required simply be a `read_all` and `write_all` with a default implementation for `mpsse_xfer`?  A lot of IO traits I see abstract only the `read`/`write` portions.  I am ok with it either way.

In addition, should some of the methods from `FtdiMpsse` be moved into this trait?  For example `clock_data` only needs `mpsse_send` and `mpsse_xfer` so it could be implemented with a default method in this trait.
https://github.com/newAM/libftd2xx-rs/blob/e3e7eaeb9a603d92a257bf61314a4137318b1c8a/src/mpsse.rs#L421-L429

> New RFC commit has been added to MR:
> 
>     * embedding Ftdi into Ftx232h
> 
>     * implement Drop for Ftdi
> 
> 
> Now the above example works fine w/o additional close operation. So `mpsse_close` is not needed as well.
> If this or similar change will be done, then `mpsse_close` can be removed from MpsseCmdExecutor trait.

That works for me!

Comment (geomatsi) [reply src/lib.rs]: Ok, thanks for clarification. That looks reasonable. So I updated MpsseCmdExecutor trait definition in ```ftdi-mpsse```. Now there are 3 basic methods: ```init```/```send```/```recv```. Other methods, if needed, can be derived from ```send```/```recv```. For instance, default implementation of ```xfer``` method has been added. Appropriate change has been also added to this PR.

Comment (geomatsi) [reply src/lib.rs]: As you mentioned, some other basic methods can be moved to ```MpsseCmdExecutor``` trait default implementations. It looks like there are two consistent options: either almost all those methods (clock, loopback, gpio, clock_data) should be added to the trait or none of them. Meanwhile often those commands are combined into sequences. For insance  there are common patterns like appending ```send_immediate```. So for now I would suggest to keep trait as small as possible, leaving command construction and execution to the end user.

Comment (geomatsi) [reply src/lib.rs]: BTW, if you are ok with the general direction, then could you please create ```ftdi-mpsse``` empty project repository in your github workspace. I will create PR from the current development [repository](https://github.com/geomatsi/ftdi-mpsse).

Comment (newAM) [reply src/lib.rs]: This is adding up to a lot of repos, so I created a github organization: https://github.com/ftdi-rs
I sent you an invite.

I think it makes sense to move `libftd2xx`, `libftd2xx-ffi`, `ftdi-mpsse`, and hal crates under this org.

Maybe we can get `libftdi1-sys` and `libftdi` under this org as well (if the maintainers agree)

Comment (newAM) [reply src/lib.rs]: I saw the invite accept, I made you an owner so that you can create repos in the organization

Comment (geomatsi) [reply src/lib.rs]: Thanks. I joined to ftdi-rs. So I will create ```ftdi-mpsse```. Later on, as soon as we are done with the reviews for this crate and ```ftdi-mpsse```, I will move ```ftdi-embedded-hal``` there as well.

Comment (newAM) [reply src/lib.rs]: > So for now I would suggest to keep trait as small as possible, leaving command construction and execution to the end user.

Sorry just read all the other comments, they were hidden under a drop down.

This sounds good to me, the traits can always be expanded with default members as a non-breaking change later.

**Comment by geomatsi** [comment]
Hi @newAM 

I created new [ftdi-mpsse](https://github.com/ftdi-rs/ftdi-mpsse) repository in the new space and posted the accompanying pull-request: https://github.com/ftdi-rs/ftdi-mpsse/pull/1. At the moment my goal is to finalize the following 3 connected PRs:
- this one
- ftdi-mpsse initial implementation: https://github.com/ftdi-rs/ftdi-mpsse/pull/1
- accompanying changes for ftd2xx-embedded-hal: https://github.com/newAM/ftd2xx-embedded-hal/pull/35

I looks like the first portion of your review comments for this RFC PR has been addressed. So it is a good time to rebase and re-organize this PR. Are you OK with force-push here ?

Regards,
Sergey

**Comment by newAM** [comment]
> Are you OK with force-push here ?

Yup, that's fine by me!

**Comment by geomatsi** [comment]
This MR and its accompanying MRs have been updated according to review comments:
- ```ftdi-mpsse```: https://github.com/ftdi-rs/ftdi-mpsse/pull/1
- ```ftd2xx-embedded-hal```: https://github.com/newAM/ftd2xx-embedded-hal/pull/35

The last MR has been simplified after the new crate definitions have been re-exported by ```libftd2xx-rs```.  I think these three MRs are cleaned-up and ready for the next review round.

After we are done with review, I think the upcoming activities can be ordered as follows:
- merge [ftdi-rs/ftdi-mpsse#1](https://github.com/ftdi-rs/ftdi-mpsse/pull/1) and publish 0.0.1 of the new crate after setting next version for ```libftd2xx-rs``` in its dev-dependencies
- merge this MR, move repository to ftdi-rs organization and then publish new version
- merge [ftd2xx-embedded-hal#35](https://github.com/newAM/ftd2xx-embedded-hal/pull/35) , move repository to ftdi-rs organization and then and publish next version after setting next version for ```libftd2xx-rs``` in its dependencies

Thoughts ? Comments ?

Regards,
Sergey

**Comment by newAM** [comment]
Work has me busy again, I will be sure to take a look this weekend!

**Comment by newAM** [comment]
I opened a PR here to add CI the `ftdi-mpsse`: https://github.com/geomatsi/ftdi-mpsse-1/pull/1
Pending your feedback on that I think it is ready for a `v0.1.0` release.

> merge ftd2xx-embedded-hal#35 , move repository to ftdi-rs organization and then and publish next version after setting next version for libftd2xx-rs in its dependencies

Is the plan to make that crate generic for `libftdi` and `libftd2xx`?  I was thinking your original name (`ftdi-embedded-hal`) would be better for the generic HAL.  Then I can just archive `fd2xx-embedded-hal` when the generic HAL is released.

**Comment by geomatsi** [comment]
> I opened a PR here to add CI the `ftdi-mpsse`: [geomatsi/ftdi-mpsse-1#1](https://github.com/geomatsi/ftdi-mpsse-1/pull/1)
> Pending your feedback on that I think it is ready for a `v0.1.0` release.

Great ! All merged. One more PR is submitted: updated list of authors and fixed CI.

> > merge ftd2xx-embedded-hal#35 , move repository to ftdi-rs organization and then and publish next version after setting next version for libftd2xx-rs in its dependencies
> 
> Is the plan to make that crate generic for `libftdi` and `libftd2xx`? I was thinking your original name (`ftdi-embedded-hal`) would be better for the generic HAL. Then I can just archive `fd2xx-embedded-hal` when the generic HAL is released.

Initially I have been experimenting with my `ftdi-embedded-hal` adding support for both `libftdi-rs` and `libftd2xx-rs` backends. Meanwhile your HAL implementation is more mature. So I suggest to copy it to  the `ftdi-rs` organization under new name `ftdi-embedded-hal`. Then we can work on making it generic for both backends.

Regards,
Sergey

**Comment by geomatsi** [comment]
BTW, do you mind if I create copies of `libftd2xx-rs` and `ftd2xx-embedded-hal` (renamed to `ftdi-embedded-hal`) in `ftd-rs` space ?

**Comment by newAM** [comment]
> Initially I have been experimenting with my ftdi-embedded-hal adding support for both libftdi-rs and libftd2xx-rs backends. Meanwhile your HAL implementation is more mature. So I suggest to copy it to the ftdi-rs organization under new name ftdi-embedded-hal. Then we can work on making it generic for both backends.

Sounds good to me.

> BTW, do you mind if I create copies of libftd2xx-rs and ftd2xx-embedded-hal (renamed to ftdi-embedded-hal) in ftd-rs space ?

For `libftd2xx-rs` I will move this repo in when this is merged.

**Comment by geomatsi** [comment]
Ok. I updated the link to `ftdi-mpsse`. It is still git repo until ftdi-mpsse is not published.  Let me know what you expect from me before you are ready to get this PR merged.

**Comment by newAM** [comment]
Ideally a release of `ftdi-mpsse` can be made before this gets merged.
Do you have a crates.io account?  I couldn't find your name there.

**Comment by newAM** [comment]
I found out that moving repos on github is actually very non-destructive and wont impact this PR so I moved the repo over to `ftdi-rs`.

**Comment by geomatsi** [comment]
> Ideally a release of `ftdi-mpsse` can be made before this gets merged.
> Do you have a crates.io account? I couldn't find your name there.

No, not yet. So if you have time to publish it, that would be great. If not, then I will register on crates.io and learn how to do it.

**Comment by newAM** [comment]
> No, not yet. So if you have time to publish it, that would be great. If not, then I will register on crates.io and learn how to do it.

Sounds good, I will do that now.

**Comment by newAM** [comment]
https://github.com/ftdi-rs/ftdi-mpsse/actions/runs/1275624692

The release is in the pipeline now :tada: 

If/when you get a `crates.io` account let me know and I will add you as an owner of the repos in ftdi-rs.

**Comment by geomatsi** [comment]
Force-pushed after updating the first commit in the PR: added ftdi-mpsse version "^0.1".

**Review by newAM — APPROVED** [review]

**Pull request merged**
